packages/labs/src/tagInput/tagInput.tsx

```diff
@@ -0,0 +1,137 @@
+}
+
+export interface ITagInputState {
+    inputValue?: string;
+    isInputFocused?: boolean;
+}
+
+@PureRender
+export class TagInput extends AbstractComponent<ITagInputProps, ITagInputState> {
+    public static displayName = "Blueprint.TagInput";
+
+    public static defaultProps: Partial<ITagInputProps> = {
+        inputProps: {},
+        tagProps: {},
+    };
+
+    public state: ITagInputState = { inputValue: "", isInputFocused: false };
+
+    private inputElement: HTMLInputElement;
+    private refHandlers = {
+        input: (ref: HTMLInputElement) => {
+            this.inputElement = ref;
+            const refHandler = this.props.inputProps.ref;
+            if (Utils.isFunction(refHandler)) {
+                refHandler(ref);
+            }
+        },
+    };
+
+    public render() {
+        const { inputProps, tagProps, values } = this.props;
+
+        const classes = classNames(Classes.INPUT, "pt-tag-input", {
+            [Classes.ACTIVE]: this.state.isInputFocused,
+        }, this.props.className);
+
+        const tags = values.map((tag, i) => {
+            const props = Utils.isFunction(tagProps) ? tagProps(tag, i) : tagProps;
+            return <Tag data-tag-index={i} key={tag + "__" + i} onRemove={this.handleRemoveTag} {...props}>{tag}</Tag>;
+        });
+        return (
+            <div className={classes} onClick={this.handleContainerClick}>
+                {tags}
+                <input
+                    value={this.state.inputValue}
+                    {...inputProps}
+                    onBlur={this.handleInputBlur}
+                    onFocus={this.handleInputFocus}
+                    onChange={this.handleInputChange}
+                    onKeyDown={this.handleInputKeyDown}
+                    ref={this.refHandlers.input}
+                    className={classNames("pt-input-invisible", inputProps.className)}
+                />
+            </div>
+        );
+    }
+
+    private handleContainerClick = () => {
+        if (this.inputElement != null) {
+            this.inputElement.focus();
+        }
+    }
+
+    private handleInputBlur = (event: React.FocusEvent<HTMLInputElement>) => {
+        this.setState({ isInputFocused: false });
+        Utils.safeInvoke(this.props.inputProps.onBlur, event);
+    }
+
+    private handleInputFocus = (event: React.FocusEvent<HTMLInputElement>) => {
+        this.setState({ isInputFocused: true });
+        Utils.safeInvoke(this.props.inputProps.onFocus, event);
+    }
+
+    private handleInputChange = (event: React.KeyboardEvent<HTMLInputElement>) => {
+        this.setState({ inputValue: event.currentTarget.value });
+        Utils.safeInvoke(this.props.inputProps.onChange, event);
+    }
+
+    private handleInputKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
+        const { value } = event.currentTarget;
+        if (event.which === Keys.ENTER && value.length > 0) {
+            Utils.safeInvoke(this.props.onAdd, value);
```

Comment:
Elsewhere, we `setState` before invoking the callback. Is the switch here intentional?